core/src/replay_stage.rs

```diff
@@ -525,34 +527,19 @@ impl ReplayStage {
         }
     }
 
-    // Returns Some(result) if the `result` is a fatal error, which is an error that will cause a
-    // bank to be marked as dead/corrupted
-    fn is_replay_result_fatal(result: &Result<()>) -> bool {
-        match result {
-            Err(Error::TransactionError(e)) => {
-                // Transactions withand transaction errors mean this fork is bogus
-                let tx_error = Err(e.clone());
-                !Bank::can_commit(&tx_error)
-            }
-            Err(Error::BlockError(_)) => true,
-            Err(Error::BlocktreeError(BlocktreeError::InvalidShredData(_))) => true,
-            Err(Error::BlocktreeError(BlocktreeError::DeadSlot)) => true,
-            _ => false,
-        }
-    }
-
     // Returns the replay result and the number of replayed transactions
     fn replay_blocktree_into_bank(
         bank: &Arc<Bank>,
         blocktree: &Blocktree,
         bank_progress: &mut ForkProgress,
         transaction_status_sender: Option<TransactionStatusSender>,
         verify_recyclers: &VerifyRecyclers,
-    ) -> (Result<()>, usize) {
+    ) -> (result::Result<(), BlocktreeProcessorError>, usize) {
         let mut tx_count = 0;
         let now = Instant::now();
         let load_result =
-            Self::load_blocktree_entries_with_shred_info(bank, blocktree, bank_progress);
+            Self::load_blocktree_entries_with_shred_info(bank, blocktree, bank_progress)
```

Comment:
@jstarry 

I'm thinking we also delete this and move the inner `blocktree::get_slot_entries_with_shred_info()` that it calls into the new consolidated `blocktree_processor::verify_and_process_entries`, and:

1) Have `blocktree_processor::verify_and_process_entries` take a start_index to start fetching shreds from (so that it can pass this to blocktree::get_slot_entries_with_shred_info()) 

2)`blocktree_processor::verify_and_process_entries` outputs the timing results below in some "VerifyAndProcessTime" struct and also bubbles up the necessary progress data `(Vec<Entry>, usize, bool)`, from the call to `blocktree::get_slot_entries_with_shred_info()`

The reasoning for the above move is that I'm beginning to dislike this `is_result_fatal` model we have here where we don't know whether to freeze a bank after an error has occurred (addressing your comments below). Instead, I think it would be cleaner if `blocktree_processor::verify_and_process_entries()` only returned fatal errors, such that if any error is returned we mark the slot as dead, otherwise the slot should be safe to continue processing. In that vein, it's nice to have the `load_blocktree` functionality also in that `blocktree_processor::verify_and_process_entries()` function so we have consistent error handling for any errors while fetching shreds from blocktree.

What do you think?

Reply:
As to where errors can bubble up:

- I think if blocktree::get_slot_entries_with_shred_info().unwrap() returns any error we can bubble it up. The only error I'm concerned about here is if the database gets corrupted for that slot, and this function errors, do we want to mark the slot as dead? I'm guessing at that point you would have to clear that part of your ledger anyways...
- Of course we bubble up all the `BlockError`'s related to tick verification
- In the actual transaction processing in `process_entries_with_callback`, the possible errors: 

a) lock results: `https://github.com/solana-labs/solana/blob/master/ledger/src/blocktree_processor.rs#L211` which are always fatal, so should be ok to propagate up

b) https://github.com/solana-labs/solana/blob/master/ledger/src/blocktree_processor.rs#L101, I think only fatal `TransactionErrors` are bubbled up here @CriesofCarrots, I think InstructionErrors are mapped away here:  https://github.com/solana-labs/solana/blob/master/runtime/src/bank.rs#L1336. In fact we can add an `assert_matches` at https://github.com/solana-labs/solana/blob/master/ledger/src/blocktree_processor.rs#L101 that it matches a TransactionError

Reply:
Sounds great to me 👍

Reply:
@carllin 

> The only error I'm concerned about here is if the database gets corrupted for that slot, and this function errors, do we want to mark the slot as dead? I'm guessing at that point you would have to clear that part of your ledger anyways...

Is it possible to recover from an "InvalidShredData" error without restarting a node?

Reply:
@jstarry because we validated the signature of the shred before we stored it into blocktree, `InvalidShredData` means either:

1) The leader signed and sent us bad data that cannot be deserialized into entries
2) Database was corrupted and the data is corrupted

Differentiating between the two would require re-validating the signature (which is also stored). In either case, I think marking the slot as dead for now is fine? The alternative is we actually try to differentiate between those two cases and drop the corrupted blobs if we believe 2) to be the case of the error, but that would also require making sure that the `SlotMeta` is updated correctly and that repair works to recover the shred.

Reply:
Seems worth having an issue for better handling for 2) because it would be pretty opaque to a validator why they forked if it was due to data corruption, right?